LinkText and VisitedText system colors doesn't obey color scheme

lukewarlow opened the pull request:
#### b538c70c4d25f6ed285f69728f78a54cb4b9ad14
<pre>
LinkText and VisitedText system colors doesn&apos;t obey color scheme
<a href="https://bugs.webkit.org/show_bug.cgi?id=272299">https://bugs.webkit.org/show_bug.cgi?id=272299</a>

Reviewed by Tim Nguyen.

Change LinkText and VisitedText CSS system colors to use a different value when the used color scheme is dark.
The chosen colors are the same as `-webkit-link` and are what is expected by WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-consistency-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-support-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-color/system-color-support-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-color/system-color-support-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-color/system-color-consistency-expected.txt:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::systemColor const):

Canonical link: <a href="https://commits.webkit.org/277259@main">https://commits.webkit.org/277259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/228bd5c55465cfe9e60bd9a487d0857ba54a959d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49798 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43164 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49422 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31573 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23751 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38380 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23778 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40605 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19687 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21176 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41774 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5159 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43502 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42161 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51674 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22139 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18515 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45674 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23416 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44679 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10396 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24198 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23133 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1559,7 +1559,7 @@ Color RenderTheme::systemColor(CSSValueID cssValueId, OptionSet<StyleColorOption
     // Text in non-active, non-visited links. For light backgrounds, traditionally blue.
     case CSSValueLinktext:
-        return SRGBA<uint8_t> { 0, 0, 238 };
+        return useDarkAppearance ? SRGBA<uint8_t> { 158, 158, 255 } : SRGBA<uint8_t> { 0, 0, 238 };
```
Can we change `CSSValueWebkitLink` to avoid duplication?

```
case CSSValueWebkitLink:
    if (forVisitedLink)
        return useDarkAppearance ? SRGBA<uint8_t> { 208, 173, 240 } : SRGBA<uint8_t> { 85, 26, 139 };
    return systemColor(CSSValueLinktext, options);
```

nt1m replied:
Actually what would be even better is that we also fix `VisitedText` to be dark theme friendly and that we do this:

```
case CSSValueWebkitLink:
    if (forVisitedLink)
        return systemColor(CSSValueVisitedtext, options);
    return systemColor(CSSValueLinktext, options);
```

lukewarlow replied:
Yeah no worries! Wasn't sure on the story behind WebKit link so didn't want to touch it. Likewise was going to do visited in a follow up but happy to do it here :)

pxlcoder commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -465,9 +465,9 @@ static Color activeButtonTextColor()
-    } else if (forVisitedLink && cssValueID == CSSValueWebkitLink) {
-        // The system color cache below can't handle visited links. The only color value
-        // that cares about visited links is CSSValueWebkitLink, so handle it here.
+    } else if (forVisitedLink && (cssValueID == CSSValueWebkitLink || cssValueID == CSSValueVisitedtext)) {
```
Is the check for `CSSValueVisitedtext` really necessary here? Its value is not actually dependent on `forVisitedLink`, so I would have expected that it's cacheable.

nt1m replied:
It doesn't look necessary from what I can understand

lukewarlow replied:
Without it the assertion after it fails because the WebKit link is reliant on this value under the hood now.

nt1m replied:
Might be better to change the assertion below so that CSSValueVisitedtext can still be cached:

```
// -webkit-link requests Visitedtext when the link is visited, let it fetch Visitedtext from the cache if possible.
ASSERT(!forVisitedLink || cssValueID == CSSValueVisitedtext);
```

(Although right now Linktext & VisitedText are never cached due to not using a system color, there is a bug to change that)

nt1m replied:
Maybe @pxlcoder has a better idea?

lukewarlow replied:
I've gone ahead with that change. Seems as reasonable as any other approach.

pxlcoder replied:
I would have probably factored out the code in `RenderTheme` into static methods, rather than making `-webkit-link` explicitly dependent on `visitedtext` – but this seems fine too.

lukewarlow committed:
LinkText and VisitedText system colors doesn't obey color scheme
https://bugs.webkit.org/show_bug.cgi?id=272299

Reviewed by Tim Nguyen.

Change LinkText and VisitedText CSS system colors to use a different value when the used color scheme is dark.
The chosen colors are the same as `-webkit-link` and are what is expected by WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-consistency-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-support-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-color/system-color-support-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-color/system-color-support-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-color/system-color-consistency-expected.txt:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::systemColor const):

Canonical link: https://commits.webkit.org/277259@main
